ragna/assistants/_cohere.py

```diff
@@ -40,23 +76,35 @@ async def answer(
                 "authorization": f"Bearer {self._api_key}",
             },
             json={
-                "preamble_override": self._make_preamble(),
-                "message": prompt,
+                "preamble_override": system_prompt,
+                "message": self._render_prompt(prompt),
```

Comment:
While the message indeed can only be a single string here, the endpoint has a [`chat_history`](https://docs.cohere.com/reference/chat#request.body.chat_history) parameter. And that one takes the previous message similar to all other assistants.

I would let `_render_prompt` return a tuple of the chat history and the current user message, e.g.

```python
chat_history, message = self._render_prompt(prompt)
```

Reply:
Conflicted on this one - seems like this would fit specifically with the pre-process pass, or this puts this one specific assistant ahead of the others in terms of capabilities - it certainly doesn't hurt anything, so happy to do so here, but also see how we might want to see what a pre-process stage looks like for all assistants and implement in one go.